fix: don't encode ':' or '/' as part of the canonical representation

dwalluck opened the pull request:
This makes the Java canonical representation match the majority of other implementations.

Fixes #122
Fixes #92

dwalluck commented:
Note that this matches the current test suite data:

https://github.com/package-url/purl-spec/blob/main/test-suite-data.json#L88-L89

https://github.com/package-url/purl-spec/blob/main/test-suite-data.json#L112-L113

jeremylong commented:
The line 112-113 where you say the `:` is encode - I don't think it is. The URL just doesn't have a schema; just a host and path: `repository_url=repo.spring.io/release`

dwalluck commented:
Let me run the full test suite again without the `':'` and see if it fails.

dwalluck commented:
OK, the first line has is for `':'`, the second line is for `'/'` (`"%2F"`) (not  `':'`).

dwalluck commented:
Checking the latest test suite file, if you don't encode the colon, if fails. The file seems to be inconsistent. It's encoded in one spot and not in the other.

```
org.junit.ComparisonFailure: 
Expected :pkg:huggingface/microsoft/deberta-v3-base@559062ad13d311b87b2c455e67dcd5f1c8f65111?repository_url=https://hub-ci.huggingface.co
Actual   :pkg:huggingface/microsoft/deberta-v3-base@559062ad13d311b87b2c455e67dcd5f1c8f65111?repository_url=https%3A//hub-ci.huggingface.co
```

jeremylong commented:
I was talking about the one you mentioned above - there doesn't seem to be any encoding.

https://github.com/package-url/purl-spec/blob/8040ff0be50f0c5b1986b1a0947bd539f5405fc4/test-suite-data.json#L112-L113
```json
    "purl": "pkg:Maven/org.apache.xmlgraphics/batik-anim@1.9.1?classifier=sources&repositorY_url=repo.spring.io/release",
    "canonical_purl": "pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1?classifier=sources&repository_url=repo.spring.io/release",
```

dwalluck commented:
There is no encoding in the input. However, the _output_ without the patch is `"%2F"`, not `'/'`, I think.

dwalluck commented:
And I found the inconsistency in the test file: the colon is encoded when it's in the version (`@sha256%3A`). It's not encoded when it's in a key value (`repository_url=https%3A`).

This is not right, is it? I am trying to verify if the problem is the code or the test file.

dwalluck commented:
> The line 112-113 where you say the `:` is encode - I don't think it is. The URL just doesn't have a schema; just a host and path: `repository_url=repo.spring.io/release`

Sorry, I flipped the examples! L112-113 is for `'/'`, and L88-89 is for `':'`. I hope it makes sense now

ppkarwasz commented on src/main/java/com/github/packageurl/PackageURL.java:
```diff
@@ -460,7 +460,7 @@ private static String uriEncode(String source, Charset charset) {
 
     private static boolean isUnreserved(int c) {
-        return (isAlpha(c) || isDigit(c) || '-' == c || '.' == c || '_' == c || '~' == c);
+        return (isAlpha(c) || isDigit(c) || '-' == c || '.' == c || '_' == c || '~' == c || ':' == c || '/' == c);
```
There should be rather a different set for each PURL component:

- `namespace` does not need to encode neither `:` nor `/`.
- `name` does not need to encode `:`, but it needs to encode `/`, which is used to separate it from `namespace`.
- `qualifiers`, probably does not need to encode `:`, `/` or `?`.

dwalluck replied:
I will add a param to the `percentEncode` to take a `String` of characters to allow.

dwalluck replied:
Really, there are a lot more characters potentially allowed by the RFC than we list here.

I am of the opinion right now to add just enough in order to pass the test suite.

The `'?'` is not allowed unencoded per the purl spec itself.

dwalluck commented:
So, with [8ab171f](https://github.com/package-url/packageurl-java/pull/161/commits/8ab171f7dc51182c63d013885cbada2f066ea19e), this now fixes #122, but _not_ #92. The test suite has `"%3A"`, but all examples given in https://github.com/package-url/purl-spec/blob/8040ff0be50f0c5b1986b1a0947bd539f5405fc4/README.rst?plain=1#L116-L117 say otherwise.

Should we just change the test suite back to how to was and assume that there is a mistake in the tests and that the current `README.rst` is the correct one?

ppkarwasz commented:
The exact definition of `namespace`, `name`, `version` and `subpath` are still WIP: https://github.com/orgs/package-url/projects/1

Maybe this PR is precocious and we should wait for the spec to be finalized.

dwalluck commented:
The author of the Rust version has proposed the set of characters to use https://github.com/phylum-dev/purl/blob/151168733f75a9802556e4b07eb577b9d99f7cea/purl/src/format.rs#L9-L27

I could take them from here.

dwalluck commented:
I will revisit this later. It may need to be combined with #174.

ppkarwasz commented:
Now that Spotless is there, there will be some conflicts. This should help:

```
git checkout origin/master -- pom.xml
mvn spotless:apply
```

dwalluck commented:
> git checkout origin/master -- pom.xml
> mvn spotless:apply

OK, it does apply on build, but if you have to fix the conflicts _first_, then it doesn't really help.

ppkarwasz commented:
> > git checkout origin/master -- pom.xml
> > mvn spotless:apply
> 
> OK, it does apply on build, but if you have to fix the conflicts _first_, then it doesn't really help.

You can:

1. Checkout `pom.xml` from master to have the Spotless configuration.
2. Format your code.
3. Merge the `master` branch into your PR branch. 99% of the code should be identical and identically formatted, so few conflicts should occur.

dwalluck committed:
fix: don't encode ':' or '/' as part of the canonical representation

This makes the Java canonical representation match the majority of
other implementations.

Fixes #122
Fixes #92

dwalluck committed:
Merge remote-tracking branch 'upstream/master' into GH-122

dwalluck committed:
Merge remote-tracking branch 'upstream/master' into GH-122

dwalluck committed:
Merge remote-tracking branch 'upstream/master' into GH-122

ppkarwasz commented:
This is blocked by package-url/purl-spec#438

dwalluck committed:
Merge remote-tracking branch 'upstream/master' into GH-122

ppkarwasz commented:
While PR package-url/purl-spec#461 haven't been merged yet, it is only receiving changes in the wording, not the rules.

I think it is pretty safe to assume that the following `byte`s must be left as-is:

- alphanumeric `a` through `z`, `A` through `Z`, and `0` through `9`.
- the five `byte`s: `.`, `-`, `_`, `~` and `:`.

So I think the only change we need to apply is to add `:` to `StringUtil.UNRESERVED_CHARS` (1 LoC). What do you think?

jeremylong commented:
@ppkarwasz I agree - we need to add `:` to the `StringUtil.UNRESERVED_CHARS`.
